### ISSUE TITLE
JMX Heap monitoring issues

### ISSUE THREAD

**Issue by yuvarajan1989k** [Stackdriver/stackdriver-agent-service-configs#52]
Hi,

I received the below error.
```
Nov 24 06:41:00 *** collectd[31721]: cjni_init_native: Cannot find the API class "org.collectd.api.Collectd". Please set the correct class path using 'JVMArg "-Djava.class.path=..."'.
Nov 24 06:41:00 ** collectd[31721]: java plugin: cjni_create_jvm: cjni_init_native failed.
Nov 24 06:41:00 ** collectd[31721]: java plugin: cjni_thread_attach: cjni_create_jvm failed.
Nov 24 06:41:00 ** collectd[31721]: java plugin: Configuration block for `GenericJMX' found, but no such configuration callback has been registered. Please make sure, the `LoadPlugin' lines precede the `Plugin' blocks.
```

JMX is configured and listening in the port 1097.
   ServiceURL "service:jmx:rmi:///jndi/rmi://localhost:1097/jmxrmi"

Java version

java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.221-b11, mixed mode)

**Comment by qingling128** [comment]
@yuvarajan1989k - Can you provide your Collectd configuration and confirm `the LoadPlugin lines precede the Plugin blocks`?

**Comment by yuvarajan1989k** [comment]
# This is the monitoring configuration for standalone JVMs.
# Look for JMX_HOST and JMX_PORT to adjust your configuration file.
# Replace JMX_PORT below with the value configured in your JVM deployment.
LoadPlugin java
<Plugin "java">
  JVMARG "-Djava.class.path=/opt/stackdriver/collectd/share/collectd/java/collectd-api.jar:/opt/stackdriver/collectd/share/collectd/java/generic-jmx.jar"
  LoadPlugin "org.collectd.java.GenericJMX"
  <Plugin "GenericJMX">
        <MBean "jvm_localhost_Threading">
            ObjectName "java.lang:type=Threading"
            <Value>
                Type "gauge"
                InstancePrefix "threading-daemon_thread_count"
                Table false
                Attribute "DaemonThreadCount"
            </Value>
            <Value>
                Type "gauge"
                InstancePrefix "threading-thread_count"
                Table false
                Attribute "ThreadCount"
            </Value>
            <Value>
                Type "gauge"
                InstancePrefix "threading-peak_thread_count"
                Table false
                Attribute "PeakThreadCount"
            </Value>
        </MBean>

        <MBean "jvm_localhost_Memory">
            ObjectName "java.lang:type=Memory"
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_committed"
                Table false
                Attribute "HeapMemoryUsage.committed"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_committed"
                Table false
                Attribute "NonHeapMemoryUsage.committed"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_used"
                Table false
                Attribute "HeapMemoryUsage.used"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_used"
                Table false
                Attribute "NonHeapMemoryUsage.used"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_max"
                Table false
                Attribute "HeapMemoryUsage.max"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_max"
                Table false
                Attribute "NonHeapMemoryUsage.max"
            </Value>
        </MBean>

        <MBean "jvm_localhost_Runtime">
            ObjectName "java.lang:type=Runtime"
            <Value>
                Type "counter"
                InstancePrefix "runtime-uptime"
                Table false
                Attribute "Uptime"
            </Value>
        </MBean>

        <MBean "jvm_localhost_os">
            ObjectName "java.lang:type=OperatingSystem"
            <Value>
                Type "gauge"
                InstancePrefix "os-open_fd_count"
                Table false
                Attribute "OpenFileDescriptorCount"
            </Value>
            <Value>
                Type "counter"
                InstancePrefix "os-process_cpu_time"
                Table false
                Attribute "ProcessCpuTime"
            </Value>
        </MBean>

        <MBean "jvm_localhost_gc">
            ObjectName "java.lang:type=GarbageCollector,name=*"
            InstanceFrom "name"
            <Value>
                Type "counter"
                InstancePrefix "gc-collection_count"
                Table false
                Attribute "CollectionCount"
            </Value>
            <Value>
                Type "counter"
                InstancePrefix "gc-collection_time"
                Table false
                Attribute "CollectionTime"
            </Value>
        </MBean>

        <Connection>
            Host "localhost"
            # When using non-standard JVM configurations, replace the below with
            #ServiceURL "service:jmx:rmi:///jndi/rmi://JMX_HOST:JMX_PORT/jmxrmi"
            ServiceURL "service:jmx:rmi:///jndi/rmi://127.0.0.1:1097/jmxrmi"
            InstancePrefix "jvm"

            Collect "jvm_localhost_Threading"
            Collect "jvm_localhost_Memory"
            Collect "jvm_localhost_Runtime"
            Collect "jvm_localhost_os"
            Collect "jvm_localhost_gc"
        </Connection>
    </Plugin>
</Plugin>

LoadPlugin match_regex
LoadPlugin target_set
LoadPlugin target_replace
<Chain "GenericJMX_jvm">
    <Rule "rewrite_genericjmx_to_jvm">
        <Match regex>
            Plugin "^GenericJMX$"
            PluginInstance "^jvm.*$"
        </Match>
        <Target "replace">
            PluginInstance "jvm" ""
        </Target>
        <Target "set">
            Plugin "jvm"
        </Target>
        Target "return"
    </Rule>
</Chain>

<Chain "PreCache">
    <Rule "jump_to_GenericJMX_jvm">
        <Target "jump">
            Chain "GenericJMX_jvm"
        </Target>
    </Rule>
</Chain>
PreCacheChain "PreCache"

**Comment by yuvarajan1989k** [comment]
`# This is the monitoring configuration for standalone JVMs.
# Look for JMX_HOST and JMX_PORT to adjust your configuration file.
# Replace JMX_PORT below with the value configured in your JVM deployment.
LoadPlugin java
<Plugin "java">
  JVMARG "-Djava.class.path=/opt/stackdriver/collectd/share/collectd/java/collectd-api.jar:/opt/stackdriver/collectd/share/collectd/java/generic-jmx.jar"
  LoadPlugin "org.collectd.java.GenericJMX"
  <Plugin "GenericJMX">
        <MBean "jvm_localhost_Threading">
            ObjectName "java.lang:type=Threading"
            <Value>
                Type "gauge"
                InstancePrefix "threading-daemon_thread_count"
                Table false
                Attribute "DaemonThreadCount"
            </Value>
            <Value>
                Type "gauge"
                InstancePrefix "threading-thread_count"
                Table false
                Attribute "ThreadCount"
            </Value>
            <Value>
                Type "gauge"
                InstancePrefix "threading-peak_thread_count"
                Table false
                Attribute "PeakThreadCount"
            </Value>
        </MBean>

        <MBean "jvm_localhost_Memory">
            ObjectName "java.lang:type=Memory"
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_committed"
                Table false
                Attribute "HeapMemoryUsage.committed"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_committed"
                Table false
                Attribute "NonHeapMemoryUsage.committed"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_used"
                Table false
                Attribute "HeapMemoryUsage.used"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_used"
                Table false
                Attribute "NonHeapMemoryUsage.used"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-heap_usage_max"
                Table false
                Attribute "HeapMemoryUsage.max"
            </Value>
            <Value>
                Type "memory"
                InstancePrefix "memory-non_heap_usage_max"
                Table false
                Attribute "NonHeapMemoryUsage.max"
            </Value>
        </MBean>

        <MBean "jvm_localhost_Runtime">
            ObjectName "java.lang:type=Runtime"
            <Value>
                Type "counter"
                InstancePrefix "runtime-uptime"
                Table false
                Attribute "Uptime"
            </Value>
        </MBean>

        <MBean "jvm_localhost_os">
            ObjectName "java.lang:type=OperatingSystem"
            <Value>
                Type "gauge"
                InstancePrefix "os-open_fd_count"
                Table false
                Attribute "OpenFileDescriptorCount"
            </Value>
            <Value>
                Type "counter"
                InstancePrefix "os-process_cpu_time"
                Table false
                Attribute "ProcessCpuTime"
            </Value>
        </MBean>

        <MBean "jvm_localhost_gc">
            ObjectName "java.lang:type=GarbageCollector,name=*"
            InstanceFrom "name"
            <Value>
                Type "counter"
                InstancePrefix "gc-collection_count"
                Table false
                Attribute "CollectionCount"
            </Value>
            <Value>
                Type "counter"
                InstancePrefix "gc-collection_time"
                Table false
                Attribute "CollectionTime"
            </Value>
        </MBean>

        <Connection>
            Host "localhost"
            # When using non-standard JVM configurations, replace the below with
            #ServiceURL "service:jmx:rmi:///jndi/rmi://JMX_HOST:JMX_PORT/jmxrmi"
            ServiceURL "service:jmx:rmi:///jndi/rmi://127.0.0.1:1097/jmxrmi"
            InstancePrefix "jvm"

            Collect "jvm_localhost_Threading"
            Collect "jvm_localhost_Memory"
            Collect "jvm_localhost_Runtime"
            Collect "jvm_localhost_os"
            Collect "jvm_localhost_gc"
        </Connection>
    </Plugin>
</Plugin>

LoadPlugin match_regex
LoadPlugin target_set
LoadPlugin target_replace
<Chain "GenericJMX_jvm">
    <Rule "rewrite_genericjmx_to_jvm">
        <Match regex>
            Plugin "^GenericJMX$"
            PluginInstance "^jvm.*$"
        </Match>
        <Target "replace">
            PluginInstance "jvm" ""
        </Target>
        <Target "set">
            Plugin "jvm"
        </Target>
        Target "return"
    </Rule>
</Chain>

<Chain "PreCache">
    <Rule "jump_to_GenericJMX_jvm">
        <Target "jump">
            Chain "GenericJMX_jvm"
        </Target>
    </Rule>
</Chain>
PreCacheChain "PreCache"`

**Comment by yuvarajan1989k** [comment]
[jvm-sun-hotspot.txt](https://github.com/Stackdriver/stackdriver-agent-service-configs/files/3894411/jvm-sun-hotspot.txt)

**Comment by yuvarajan1989k** [comment]
Thanks for your help.

**Comment by igorpeshansky** [comment]
@yuvarajan1989k, what version of the Stackdriver monitoring agent package are you running? On what OS?
Can you please also list the contents of the `/opt/stackdriver/collectd/share/collectd/java` directory on your system?

**Comment by yuvarajan1989k** [comment]
Hi,

OS Version:- Ubuntu 18.04
stackdriver-agent: 5.5.2-1001.bionic  

ls -tlr /opt/stackdriver/collectd/share/collectd/java

-rw-r--r-- 1 root root 16759 Nov 14 11:08 generic-jmx.jar
-rw-r--r-- 1 root root 13737 Nov 14 11:08 collectd-api.jar

**Comment by igorpeshansky** [comment]
Interesting. The files are right there, and added to the classpath, but not visible to the JVM.
Is the agent running at root?
Are you seeing any errors in the logs prior to "Cannot find the API class"? E.g., "class file has the wrong version" or "bad class file"?

**Comment by yuvarajan1989k** [comment]
root     19400     1  0 Nov26 ?        00:00:30 /opt/stackdriver/collectd/sbin/stackdriver-collectd -C /etc/stackdriver/collectd.conf -P /var/run/stackdriver-agent.pid

FYI:-

Nov 26 23:37:31 ** collectd[19393]: plugin_load: plugin "java" successfully loaded.
Nov 26 23:37:31 ** collectd[19399]: plugin_load: plugin "java" successfully loaded.
Nov 26 23:37:31 ** collectd[19400]: cjni_init_native: Cannot find the API class "org.collectd.api.Collectd". Please set the correct class path using 'JVMArg "-Djava.class.path=..."'.
Nov 26 23:37:31 ** collectd[19400]: java plugin: cjni_create_jvm: cjni_init_native failed.
Nov 26 23:37:31 ** collectd[19400]: java plugin: cjni_thread_attach: cjni_create_jvm failed.
Nov 26 23:37:31 ** collectd[19400]: java plugin: Configuration block for `GenericJMX' found, but no such configuration callback has been registered. Please make sure, the `LoadPlugin' lines precede the `Plugin' blocks.

JAVA VERSION:-

java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.221-b11, mixed mode)

**Comment by igorpeshansky** [comment]
@sophieyfang, you were just (successfully) testing JVM monitoring on Bionic, right? Can you please compare your environment with the above? Were you also using JRE 1.8? I think you may have tested with the full JDK, though it shouldn't affect the ability to load the plugin classes. Any other discrepancies?

**Comment by sophieyfang** [comment]
I just tested it again on ubuntu1804 with jdk8 and jdk11. Previously, i was only testing it on jdk11 as it's latest one supported on ubuntu1804. jdk 11 has no issues and i can see the jvm metrics flow in. However, jdk8 does give the below error:
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8991]: plugin_load: plugin "target_set" successfully loaded.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8991]: plugin_load: plugin "target_replace" successfully loaded.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 stackdriver-agent[8951]:    ...done.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 systemd[1]: Started LSB: start and stop Stackdriver Agent.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: cjni_init_native: Cannot find the API class "org.collectd.api.Collectd". Please set the correct class path using 'JVMArg "-Djava.class.path=..."
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: java plugin: cjni_create_jvm: cjni_init_native failed.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: java plugin: cjni_thread_attach: cjni_create_jvm failed.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: java plugin: Configuration block for `GenericJMX' found, but no such configuration callback has been registered. Please make sure, the `LoadPlug
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: Initialization complete, entering read-loop.
Nov 27 04:17:43 sophieyfang-ubuntu1804-jvm8 collectd[8992]: tcpconns plugin: Reading from netlink succeeded. Will use the netlink method from now on.

$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

**Comment by yuvarajan1989k** [comment]
@sophieyfang  is it possible to run this with java 1.8 ?

**Comment by igorpeshansky** [comment]
@yuvarajan1989k, thanks for the report. We should support all JDK versions available from the default repos on a given distro (presumably including JDK 8 on Bionic). As you can see, we've reproduced your issue, and will work on fixing it.

**Comment by igorpeshansky** [comment]
@yuvarajan1989k In the meantime, you could try the following workaround:
- install JDK 11, but don't make it the default.
- add the location of that JDK's `libjvm.so` as `JAVA_LIB_DIR` to `/etc/default/stackdriver-agent`, e.g.:
  ```
  JAVA_LIB_DIR="/usr/lib/jvm/java-11-openjdk-amd64/jre/lib/amd64/server"
  ```
- restart the agent

Please let us know if it works for you.

**Comment by yuvarajan1989k** [comment]
Yes. Above solution has helped me to solve the issues. thank you.

**Comment by igorpeshansky** [comment]
This is actually fixed by Stackdriver/collectd#183 (and Stackdriver/collectd#184 for the older agents).

**Comment by igorpeshansky** [comment]
The fix has been released in version [6.0.3](https://github.com/Stackdriver/agent-packaging/releases/tag/6.0.3) of the Stackdriver monitoring agent.
